k2/csrc/fsa_utils_inl.h

```diff
@@ -203,26 +199,228 @@ Array1<FloatType> GetForwardScores(FsaVec &fsas, Ragged<int32_t> &state_batches,
+  }
+
+  // compute leaving arc numbers in each batch
+  Array1<int32_t> leaving_arc_nums(c, num_batches);
+  Array1<int32_t> leaving_arc_start_index(c, num_batches);
+  int32_t *leaving_arc_nums_data = leaving_arc_nums.Data();
+  int32_t *leaving_arc_start_index_data = leaving_arc_start_index.Data();
+  const int32_t *fsa_in_batch_row_splits =
+      leaving_arc_batches.RowSplits(1).Data();
+  const int32_t *state_in_batch_row_splits =
+      leaving_arc_batches.RowSplits(2).Data();
+  const int32_t *leaving_arc_row_splits =
+      leaving_arc_batches.RowSplits(3).Data();
+  auto lambda_set_leaving_arc_nums =
+      [=] __host__ __device__(int32_t batch_idx) {
+        int32_t first_state_idx_in_batch =
+                    state_in_batch_row_splits[batch_idx * num_fsas],
+                next_last_state_idx_in_batch =
+                    state_in_batch_row_splits[(batch_idx + 1) * num_fsas];
+        int32_t first_leaving_arc_idx =
+                    leaving_arc_row_splits[first_state_idx_in_batch],
+                next_last_leaving_arc_idx =
+                    leaving_arc_row_splits[next_last_state_idx_in_batch];
+        leaving_arc_start_index_data[batch_idx] = first_leaving_arc_idx;
+        leaving_arc_nums_data[batch_idx] =
+            next_last_leaving_arc_idx - first_leaving_arc_idx;
+      };
+  Eval(c, num_batches, lambda_set_leaving_arc_nums);
+
+  const int32_t *fsa_in_batch_row_ids = leaving_arc_batches.RowIds(1).Data();
+  const int32_t *state_in_batch_row_ids = leaving_arc_batches.RowIds(2).Data();
+  const int32_t *leaving_arc_row_ids = leaving_arc_batches.RowIds(3).Data();
+  const int32_t *leaving_arc_ids = leaving_arc_batches.values.Data();
+  const int32_t *states_data = state_batches.values.Data();
+  const Arc *arcs = fsas.values.Data();
+  Array1<FloatType> leaving_arc_score_values(
+      c, num_arcs);  // leaving arc_scores in batches
+  FloatType *arc_scores_data = leaving_arc_score_values.Data();
+  // convert leaving_arc_nums to cpu as we will access its elements in below
+  // Eval function for `lambda_set_leaving_arc_scores`
+  leaving_arc_nums = leaving_arc_nums.To(GetCpuContext());
+  const int32_t *cpu_leaving_arc_nums = leaving_arc_nums.Data();
+  Array1<int32_t> cpu_leaving_arc_start_index =
+      leaving_arc_start_index.To(GetCpuContext());
+  const int32_t *cpu_leaving_arc_start = cpu_leaving_arc_start_index.Data();
+  // copy row_splits2 to cpu
+  Array1<int32_t> cpu_row_splits2 =
+      leaving_arc_batches.RowSplits(2).To(GetCpuContext());
+  const int32_t *cpu_row_splits2_data = cpu_row_splits2.Data();
+  Array1<int32_t> arc_sum_row_splits(c, num_states + 1);
+  Array1<FloatType> score_cache(c, num_states + 1);
+  // process batch sequentially.
+  for (int32_t i = num_batches - 1; i >= 0; --i) {
+    // get the range we would call Max/LogSum per sub list
+    int32_t row_begin =
+        cpu_row_splits2_data[i * num_fsas];  // row_start is idx01x
```

Comment:
is this even different from cpu_leaving_arc_start_index?

Reply:
... and row_end isn't a great name, I'd prefer something like this_leaving_arc_idx0xx and next_leaving_arc_idx0xx, assuming that's acurate.

Reply:
`cpu_row_splits2` is for state idx in batch (as I noted, row_begin is `idx01x`). `leaving_arc_start_index` is for arc idx in batch. they are on different levels.

Reply:
OK, will change the name.

Reply:
But you seem to be only accessing leaving_arc_start_index for the 1st state of each FSA, and the other elements aren't used.  FSA<-->CPU transfer is expensive!  I think this is the same information that is in cpu_leaving_arc_start[i].

Reply:
Sorry I meant you seem to be only accessing      cpu_row_splits2_data for the 1st state of each FSA.

Reply:
> Sorry I meant you seem to be only accessing cpu_row_splits2_data for the 1st state of each FSA.

I guess you mean `1st state of each batch`. 
 Yeah, that's true, so you mean I should just copy the 1st state of each batch to a cpu array?

Reply:
You could do that (as a single array obviously), but I think the information you need is exactly the info in leaving_arc_start_index, so you can just use that, it's already on CPU.

Reply:
Updated with RowSplits(2)[RowSplits(1)].To(cpu)